### DISCUSSION TITLE
Dynamic Error Fields not being rendered on 422

### DISCUSSION THREAD

**Discussion by Morzaram** [unpoly/unpoly#550] [Q&A]
Hey there, I'm using Phoenix for my app and on retuning a 422 the form errors are not being injected. With the debug I've got this

<img width="1178" alt="image" src="https://github.com/unpoly/unpoly/assets/70202379/678cae7b-7d23-4f65-bbfb-588c816bd46d">

As you can see here I have the error showing on the 422 
<img width="922" alt="image" src="https://github.com/unpoly/unpoly/assets/70202379/4d4d177f-0aec-48ff-a651-15916ee9461a">

Is there a certain attribute I need to have for this so the error fields get rendered?

Perhaps the way that I have things is breaking it since I'm having 2 forms and referencing them outside of it

To be clear I'm working with the `#convo-form` submission
```heex
<div data-controller="wizard" class="static w-full overflow-y-auto">
      <.simple_form
        id="convo-form"
        for={@form}
        action={~p"/v2/chats/new"}
        up-source={~p"/v2/chats/new"}
        up-submit="true"
        up-target="#convo-form"
        up-validate="true"
      >
      </.simple_form>
      <div data-wizard-target="page" class="pb-4">
        <.input
          field={@form[:name]}
          type="text"
          label="Conversation Name"
          form="convo-form"
          autocomplete="off"
          minlength="1"
        />
      </div>

      <.simple_form
        id="voice-form"
        for={@voice_form}
        action={~p"/v2/voices/new"}
        data-action="wizard#addVoice"
        class="flex flex-row items-end gap-2"
      >
        <.input field={@voice_form[:name]} type="text" label="Create A New Voice" autocomplete="off" />
        <.button type="submit" class="max-h-10">Add</.button>
      </.simple_form>

      <fieldset class="flex flex-col mt-5">
        <legend class="pb-2">Current Voices</legend>
        <div class="grid max-h-full grid-cols-2 gap-1 ">
          <label
            :for={v <- @current_user.voices}
            for={v.id}
            class="flex flex-row items-center w-full h-full p-2 border border-l-4 rounded-lg border-stone-100 "
            id={"voice-#{v.id}"}
            style={"border-color: #{v.color};
                border-top-color: rgb(231 229 228);
                border-right-color: rgb(231 229 228);
                border-bottom-color: rgb(231 229 228);"}
            data-wizard-target="voice"
          >
            <input
              id={v.id}
              type="checkbox"
              name="topic[voice_ids][]"
              value={v.id}
              form="convo-form"
              style="border-color: #{v.color};"
              autocomplete="off"
              class="w-4 h-4 mr-2 border-gray-300 rounded appearance-none checked:border-transparent checked:bg-gray-800 checked:hover:bg-gray-700"
            />
            <%= v.name %>
          </label>
        </div>
      </fieldset>

      <div class="absolute bottom-8 right-8 sm:hidden">
        <.button type="submit" form="convo-form">
          Save
        </.button>
      </div>

      <div class="hidden pt-5 text-right sm:block">
        <.button type="submit" form="convo-form" class="text-right">
          Save
        </.button>
      </div>
    </div>
```

Would love some insight on what I'm doing wrong. It might be due to the way that I'm working with nested forms but I'm not completly sure on this.

Love this library, thank you so much for y'alls hard work.

**Comment by Morzaram** [comment]
Okay so it turns out that if I put the input within the form it will render the errors

```html
  <.simple_form
        id="convo-form"
        for={@form}
        action={~p"/v2/chats/new"}
        up-source={~p"/v2/chats/new"}
        up-submit="true"
        up-target="#convo-form"
        up-validate="true"
      >
        <.input
          field={@form[:name]}
          type="text"
          label="Conversation Name"
          form="convo-form"
          autocomplete="off"
          minlength="1"
        />
      </.simple_form>
```

Luckily for me this is all I need but I wondering, is there's a better approach for handling the nested forms now with unpoly?

**Reply by adam12** [reply]
It's too hard to tell from the markup you provided.

Are you saying that you have a ` <form>` inside a `<form>`? That's not spec compliant if so, and might explain your issues. If that's not the case, I'd need more information.

**Reply by Morzaram** [reply]
Yea basically I'm using 

```html
<div>
  <form id="convo-form"></form>
  <form id="voice-form"></form>
  <div class="input-with-dynamic-errors">
    <input form="convo-form" type="text" />
    // -- Dynamic error spans rendered here
  </div>
 
  <button type="submit" form="convo-form"> Submit </submit>
</div>

```

The issue was that due to the input not being wrapped within the form it wouldn't render the errors. But I think the rule is basically, if you need to render validations from the server, keep it within the form. 

I think it's considered resolved. Thank you, and sorry 🙈

**Reply by triskweline** [reply]
Some background here:

- When the server responds with an error code, Unpoly renders the selector in the `[up-fail-target]` attribute, not the selector in the `[up-target]` attribute.
- The default `[up-fail-target]` for a form is `form`, meaning the `<form>` element itself is going to be rendered.
- You can render elements outside the form by including it in the `[up-fail-target]`:
  ```html
  <form up-submit up-target="#success" up-fail-target="form, #errors-outside-form">
   ```

Also see [rendering failed responses differently](https://unpoly.com/failed-responses#rendering-failed-responses-differently).
